File: src/main/java/com/giffardtechnologies/restdocs/domain/FieldReference.java
package com.giffardtechnologies.restdocs.domain;

import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class FieldReference {
	final String mNode;
	@Nullable
	final FieldReference mChild;

	public static FieldReference fromParts(List<String> parts) {
		String node = parts.remove(0);
		if (parts.isEmpty()) {
			return new FieldReference(node, null);
		} else {
			return new FieldReference(node, fromParts(parts));
		}
	}
	public static FieldReference fromString(String reference) {
		List<String> parts = Arrays.stream(reference.split("\\.")).collect(Collectors.toList());
		return fromParts(parts);
	}

	public FieldReference(String node, @Nullable FieldReference child) {
		mChild = child;
		mNode = node;
	}

	public String getNode() {
		return mNode;
	}

	@Nullable
	public FieldReference getChild() {
		return mChild;
	}

	/**
	 * Tests if the current node is equal to the passed string.
	 *
	 * @param node node string to check against
	 *
	 * @return true if the current node equals passed node, false otherwise
	 */
	public boolean isNode(String node) {
		return mNode.equals(node);
	}

	/**
	 * Tests if the current node is equal to the passed string and that this is a leaf node.
	 *
	 * @param node node string to check against
	 *
	 * @return true if the current node equals passed node and is a leaf, false otherwise
	 */
	public boolean isLeafNode(String node) {
		return mNode.equals(node) && mChild == null;
	}

	/**
	 * Tests if the current node is a leaf node.
	 *
	 * @return true if the current node is a leaf, false otherwise
	 */
	public boolean isLeafNode() {
		return mChild == null;
	}

}

File: src/main/java/com/giffardtechnologies/restdocs/domain/FieldReference.kt
package com.giffardtechnologies.restdocs.domain

import java.util.*
import java.util.stream.Collectors

class FieldReference(val node: String, val child: FieldReference?) {

    /**
     * Tests if the current node is equal to the passed string.
     *
     * @param node node string to check against
     *
     * @return true if the current node equals passed node, false otherwise
     */
    fun isNode(node: String): Boolean {
        return this.node == node
    }

    /**
     * Tests if the current node is equal to the passed string and that this is a leaf node.
     *
     * @param node node string to check against
     *
     * @return true if the current node equals passed node and is a leaf, false otherwise
     */
    fun isLeafNode(node: String): Boolean {
        return this.node == node && child == null
    }

    val isLeafNode: Boolean
        /**
         * Tests if the current node is a leaf node.
         *
         * @return true if the current node is a leaf, false otherwise
         */
        get() = child == null

    companion object {
        fun fromParts(parts: MutableList<String>): FieldReference {
            val node: String = parts.removeAt(0)
            return if (parts.isEmpty()) {
                FieldReference(node, null)
            } else {
                FieldReference(
                    node,
                    fromParts(parts)
                )
            }
        }

        fun fromString(reference: String): FieldReference {
            val parts = Arrays.stream(reference.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
                .collect(Collectors.toList())
            return fromParts(parts)
        }
    }
}
